### PR TITLE
Adds msg generation support for workspace overlays

### LINKED ISSUE #902: Message generation ignores overlays
**Description**
The message generation seems to ignore workspace overlays. If you have `std_msgs` installed and a checkout from source with local changes, the source checkout should be used instead of the installed messages. It seems like the generator script just runs through all entries in `AMENT_PREFIX_PATH` and generates definitions for all of them ignoring overlays. It starts off in order of `AMENT_PREFIX_PATH` but I think each entry is done async and without specific ordering so possibly any outcome could occur?

- Library Version:
- ROS Version:
- Platform / OS:

**Steps To Reproduce**
checkout `std_msgs` locally. Make a change to it and build. `AMENT_PREFIX_PATH` will look like `/path/to/source/install/std_msgs:/opt/ros/foxy`. Generate messages through `scripts/generate_messages.js`.

Note: I haven't actually done this with `std_msgs` -it's occurring for me with a custom message package

**Expected Behavior**
Generated messages will use the message definitions from the source build.

**Actual Behavior**
Generated messages will use message definitions from the installed package

### PR DESCRIPTION
**Pull request by wayneparrott**
Revises the message_generation process as follows:
1. recursively delete the `generated/` folder before generating new message files
2. process the AMENT_PREFIX_PATH in reverse order such that interfaces defined earlier on the AMENT_PREFIX_PATH will overwrite older versions of the interface. Thus interfaces defined in a ROS overlay workspace take precedent over a version of the interface defined in an underlay.

Additional mods were made to help address a stability issue involving content-filters in CI:
- updated mocha to version 10.2
- added native clearContentFilter() to rcl_bindings.cpp
- reimplemented subscription#clearContentFilter() to use new native implementation
- edited test-subscription-content-filter.js

Fix #902

### CODE REVIEW

File: .github/workflows/identify-ros-distro.yml [hunk 1]
```diff
@@ -26,6 +26,7 @@ jobs:
         id: identify
         run: |
           if ${ROLLING_VAR} == true; then
+            # echo "::set-output name=distro::rolling"
```
Comment (minggangw) [inline .github/workflows/identify-ros-distro.yml:29]: is it still needed?

File: rosidl_gen/idl_generator.js [hunk 2]
```diff
@@ -231,6 +238,13 @@ async function generateActionJSStruct(actionInfo, dir) {
 }
 
 async function generateJSStructFromIDL(pkg, dir) {
+  //   for (let messageInfo of pkg.messages)
+  //     await generateMessageJSStruct(messageInfo, dir);
+  //   for (let serviceInfo of pkg.services)
+  //     await generateServiceJSStruct(serviceInfo, dir);
+  //   for (let actionInfo of pkg.actions)
+  //     await generateActionJSStruct(actionInfo, dir);
```
Comment (minggangw) [inline rosidl_gen/idl_generator.js]: can be removed?

File: src/rcl_bindings.cpp [hunk 3]
```diff
@@ -838,6 +842,38 @@ NAN_METHOD(SetContentFilter) {
 #endif
 }
 
+NAN_METHOD(ClearContentFilter) {
+#if ROS_VERSION < 2205  // 2205 => Humble+
+  info.GetReturnValue().Set(Nan::False());
+  return;
+#else
+  // v8::Local<v8::Context> currentContext = Nan::GetCurrentContext();
```
Comment (minggangw) [inline src/rcl_bindings.cpp]: Can be removed?

File: src/rcl_bindings.cpp [hunk 4]
```diff
@@ -838,6 +842,38 @@ NAN_METHOD(SetContentFilter) {
...
+      RCL_RET_OK,
+      rcl_subscription_content_filter_options_init(
+          subscription, "", 0, (const char**)nullptr, &options),
+      rcl_get_error_string().str);
+
+  THROW_ERROR_IF_NOT_EQUAL(
+      RCL_RET_OK, rcl_subscription_set_content_filter(subscription, &options),
+      rcl_get_error_string().str);
+
+  //  rcl_ret_t ret = rcl_subscription_set_content_filter(subscription,
```
Comment (minggangw) [inline src/rcl_bindings.cpp]: Can be removed?

File: scripts/run_test.js [hunk 5]
```diff
@@ -20,12 +20,12 @@ const os = require('os');
...
-  process.env.AMENT_PREFIX_PATH + path.delimiter + actionPath;
-let msgPath = path.join(rootDir, 'test', 'rclnodejs_test_msgs');
-process.env.AMENT_PREFIX_PATH =
-  process.env.AMENT_PREFIX_PATH + path.delimiter + msgPath;
+// let actionPath = path.join(rootDir, 'test', 'ros1_actions');
+// process.env.AMENT_PREFIX_PATH =
+//   process.env.AMENT_PREFIX_PATH + path.delimiter + actionPath;
+// let msgPath = path.join(rootDir, 'test', 'rclnodejs_test_msgs');
+// process.env.AMENT_PREFIX_PATH =
+//   process.env.AMENT_PREFIX_PATH + path.delimiter + msgPath;
```
Comment (minggangw) [inline scripts/run_test.js]: Can be removed?

File: rosidl_gen/index.js [hunk 6]
```diff
@@ -20,9 +20,20 @@ const packages = require('./packages.js');
...
+
+// async function generateInPath(path) {
+//   const pkgs = await packages.findPackagesInDirectory(path);
+//   const pkgsInfo = Array.from(pkgs.values());
+
+//   // overlay support requires msg generation to synchronous, do not Promise.all()
+//   for (let pkgInfo of pkgsInfo) {
+//     await generateJSStructFromIDL(pkgInfo, generatedRoot);
+//   }
+// }
```
Comment (minggangw) [inline rosidl_gen/index.js]: Can be removed?

**Review by minggangw — APPROVED** [review]
Thanks for the patch, LGTM!

File: .github/workflows/identify-ros-distro.yml (hunk 1 above)
Comment (wayneparrott) [reply .github/workflows/identify-ros-distro.yml:29]: my bad. was having some trouble with rolling. commented out that line and the did a silly by pasting in the same line. So no, do not need this line.

File: rosidl_gen/idl_generator.js (hunk 2 above)
Comment (wayneparrott) [reply rosidl_gen/idl_generator.js]: yes, I removed that old code.

File: src/rcl_bindings.cpp (hunk 4 above)
Comment (wayneparrott) [reply src/rcl_bindings.cpp]: yes, removed the commented line

File: scripts/run_test.js (hunk 5 above)
Comment (wayneparrott) [reply scripts/run_test.js]: yes, I removed it

File: rosidl_gen/index.js (hunk 6 above)
Comment (wayneparrott) [reply rosidl_gen/index.js]: yes, removed it

File: src/rcl_bindings.cpp (hunk 3 above)
Comment (wayneparrott) [reply src/rcl_bindings.cpp]: yes

**Comment by wayneparrott** [comment]
@minggangw Thx for the review.

**Comment by minggangw** [comment]
Thanks @wayneparrott , I'm going to merge this PR!

**Pull request merged**